File: SOLID/DIP/Compliant/LibraryServiceDI.java
interface FineCalculator {
    int calculateFine(int lateDays);
}

class RegularFineCalculator implements FineCalculator {
    public int calculateFine(int lateDays) {
        return lateDays * 5;
    }
}

class ExamTimeFineCalculator implements FineCalculator {
    public int calculateFine(int lateDays) {
        return lateDays * 2;
    }
}

class NoFineCalculator implements FineCalculator {
    public int calculateFine(int lateDays) {
        return 0;
    }
}

class LibraryService {
    private FineCalculator calculator;

    LibraryService(FineCalculator calculator) {
        this.calculator = calculator;
    }

    int getFine(int lateDays) {
        return calculator.calculateFine(lateDays);
    }
}
public class LibraryServiceDI {
    public static void main(String[] args) {

        FineCalculator calculator = new ExamTimeFineCalculator();
        LibraryService service = new LibraryService(calculator);

        System.out.println(service.getFine(3));
    }
}

File: SOLID/DIP/Compliant/LibraryServiceDI.ts
// Abstraction
interface FineCalculator {
  calculateFine(lateDays: number): number;
}

// Implementations
class RegularFineCalculator1 implements FineCalculator {
  calculateFine(lateDays: number): number {
    return lateDays * 5;
  }
}

class ExamTimeFineCalculator implements FineCalculator {
  calculateFine(lateDays: number): number {
    return lateDays * 2;
  }
}

class NoFineCalculator implements FineCalculator {
  calculateFine(lateDays: number): number {
    return 0;
  }
}

// High-level module
class LibraryService1 {
  private calculator: FineCalculator;

  constructor(calculator: FineCalculator) {
    this.calculator = calculator;
  }

  getFine(lateDays: number): number {
    return this.calculator.calculateFine(lateDays);
  }
}

// Main
const calculator: FineCalculator = new ExamTimeFineCalculator();
const service = new LibraryService1(calculator);

console.log(service.getFine(3));
